src/interval_sets.jl

```diff
@@ -1,29 +1,76 @@
+API or by passing the set to `Array`.
+
+see also: https://en.wikipedia.org/wiki/Interval_arithmetic#Interval_operators
+
+## Examples
+
+```jldoctest
+julia> using Intervals
+julia> Array(union(IntervalSet(1..5), IntervalSet(3..8)))
+1-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 8)
+
+julia> Array(intersect(IntervalSet(1..5), IntervalSet(3..8)))
+1-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(3, 5)
+ 
+julia> Array(symdiff(IntervalSet(1..5), IntervalSet(3..8)))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 3)
+ Interval{Int64, Open, Closed}(5, 8)
+
+julia> Array(union(IntervalSet([1..2, 2..5]), IntervalSet(6..7)))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 5)
+ Interval{Int64, Closed, Closed}(6, 7)
+
+julia> Array(union(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 10)
+ Interval{Int64, Closed, Closed}(12, 14)
+
+julia> Array(intersect(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(4, 5)
+ Interval{Int64, Closed, Closed}(8, 9)
+
+julia> Array(setdiff(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 4)
+ Interval{Int64, Open, Closed}(9, 10)
+```
 """
-struct IntervalSet{T<:AbstractInterval} <: AbstractSet{T}
-    # TODO: Use Dict internally once union!(Vector{AbstractInterval}) is fully deprecated
-    items::Vector{T}
+struct IntervalSet{I <: AbstractInterval, A <: AbstractVector{I}}
+    items::A
 end
 
-IntervalSet(v::AbstractVector) = IntervalSet{eltype(v)}(v)
-IntervalSet(interval::T) where T <: AbstractInterval = IntervalSet{T}([interval])
+IntervalSet{T}(intervals) where T <: AbstractInterval = IntervalSet{T, Vector{T}}(intervals)
+IntervalSet(interval::T) where T <: AbstractInterval = IntervalSet{T, Vector{T}}([interval])
 IntervalSet(interval::IntervalSet) = interval
 IntervalSet(itr) = IntervalSet{eltype(itr)}(collect(itr))
-IntervalSet() = IntervalSet{AbstractInterval}(AbstractInterval[])
+IntervalSet() = IntervalSet(AbstractInterval[])
 
-Base.copy(intervals::IntervalSet{T}) where T = IntervalSet{T}(copy(intervals.items))
+Base.copy(intervals::IntervalSet{T, A}) where {T, A} = IntervalSet{T, A}(copy(intervals.items))
 Base.length(intervals::IntervalSet) = length(intervals.items)
 Base.iterate(intervals::IntervalSet, args...) = iterate(intervals.items, args...)
 Base.eltype(::IntervalSet{T}) where T = T
-Base.:(==)(a::IntervalSet, b::IntervalSet) = issetequal(a, b)
+Base.:(==)(a::IntervalSet, b::IntervalSet) = a.items == b.items
+Base.isequal(a::IntervalSet, b::IntervalSet) = isequal(a, b)
+Base.Array(intervals::IntervalSet) = intervals.items
```

Comment:
I don't believe we want to include this. This doesn't fit any existing `Array` constructor dispatch and in most cases you probably just want to call `collect`.

Reply:
This goal is to avoid a copy. This seemed like the most natural name to access the underlying array. I've seen similar patterns used elsewhere. I guess it could also be called `parent`, but that seems like more of a pun.

Reply:
One could just access `.items`, but I think having an function helps to communicate that this *could* avoid a copy but might involve some computation (e.g. if the internal representation of `IntervalSet` changes).

Reply:
I also think this is perfectly in line with the semantics of `Array` from base. The method constructs an array from its single argument.